### ISSUE TITLE
Dead?

### ISSUE THREAD

**Issue by UnknownEvil** [evolvio/evolv.io#186]
Is this project dead?

**Comment by PRabahy** [comment]
I hope not. I've just been busy for the summer, but I still think its really cool.

**Comment by UnknownEvil** [comment]
I do too, I just wasn't seeing anything happening.

**Comment by Holkin** [comment]
I'm working on major refactoring.
Dropped already all rocks as irrelevant. Everything else left untouched. 
It might help to finish it much faster if someone knows how to make this application actually work. I see creatures only going in circles, nothing special. I don't see much sense to invest time in something that actually doesn't work.

**Comment by PRabahy** [comment]
What do you want to see? Right now there is a lot of good evolution
happening. Most of the first creatures die. Over time they learn to eat,
and eventually they learn to avoid the water. Given all of the constants,
going in circles appears to be the best behavior. If you want to change
that, give them a more complex goal. I have tweaked it so that turning is
more expensive and gotten some interesting behavior from that.

**Comment by Holkin** [comment]
That's what I'm taking about.
I have version without 'improvements' and it gave me very interesting result after about 10 hours of simulation. They actually managed to change mouth color to eat dead bodies. 
This new version with eyes and attack range does nothing except conquering one specific color and running in circles. 
Just give me settings, which provide anything new.

**Comment by Holkin** [comment]
Just as a teaser: I expect at least 10 times better performance after this refactoring. :)

**Comment by UnknownEvil** [comment]
In order to get more complex behavior, we need more complex input senses and output actions.

**Comment by PRabahy** [comment]
I eventually want to have each creature have "n" eyes, each with its own angle and distance. In the real world, predators end up with their eyes facing forward more, and prey have them wider facing.

**Comment by UnknownEvil** [comment]
But, are they going to have a field of view, or is it going to stay lines? Having an FOV would be pretty useful to the predators and prey

**Comment by DiZZ3** [comment]
Hey guys, I want to start checking this project out and contribute. Is their someone here who could release an updated/refactored code so i can get started with an up to date version?? Let me know. Thanks.

**Comment by PRabahy** [comment]
@DiZZ3 Why wait? Just jump in. There hasn't been too much public activity recently, so feel free to go for it. The version in master works just fine right now. If you make an improvement/enhancement, just open a Pull Request and we will review it.

**Comment by PRabahy** [comment]
FYI, https://github.com/evolvio/evolv.io/network shows a nice graph of what people are working on. Looks like @AtomIsHere is the only person to have pushed any code to github recently.

**Comment by PaxJaromeMalues** [comment]
I am not exactly sure if I understand all of this correctly, but so far i have the impression that many of the current commits limit this softwares ability to actually produce unexpected results, which is (afaik) a main factor in evolution in general. I am not a developer myself, but that is what I am reading from this and the previous discussions I observed. IT seems that it currently evolves _(I appreciate the slow claps for this pun)_ into being a software that emulates what we already know, instead of actually giving each creature free roam over its genetic pool.
Is there yet a process for each creature which 'allows' it to 'implement' specific genetic treats by a random chance (say due to 'will of nature') ?

**Comment by PRabahy** [comment]
@PaxJaromeMalues Ideally, this simulator would only limit creatures by the laws of nature. For example, if it takes "x" amount of energy to lift an object, it would take "x * 2" energy to lift an object with twice the mass.

Unfortunately, it is extremely hard to model completely arbitrary behaviors, so I think each creature should have list of "senses" and "actions". The sense that already are coded are sensing its own size, sensing its own color, and looking, but I would love to see additional senses added.

I already made some progress making the actions that a creature can do quite flexible. https://github.com/evolvio/evolv.io/blob/master/src/evolv/io/CreatureAction.java lists out all the possible actions that have currently been coded.

In the long run, I want to make it so that each creature can evolve to have any combination of sensors and actions. I would love if sensors and actions could be generated on the fly, but I don't know how to do that programmatically in a way that could be displayed on the screen. I have no problem adding "made up" sensor like telepathy if somebody is willing to code them up.

**Comment by NateUniAccount** [comment]
It would be a shame if this died, but imo the original release didn't do itself many favours. Just one opinion, but when I was contributing I felt there were many design choices made in the original release that have since caused issues with eg. performance, ease of understanding, ease of contributing, flexibility. Of course YMMV. I felt that the only way forward would be to either spend huge amounts of time and thought on it, or to start over. Neither was very appealing.
I still check in every now and then to see if I can help out a bit but that's about all I've got time for now really with being busy with work and other projects.

**Comment by Holkin** [comment]
For now I have about 60% CPU time handling collisions. Does anybody know some solution to improve collision detection?

**Comment by PRabahy** [comment]
@Holkin I did a little research into that and it looks like the solution is to use a https://en.wikipedia.org/wiki/Quadtree. Unfortunately, I don't really understand exactly how to construct one, but the tree based searching makes a lot of sense as to why it would be faster.

**Comment by NateUniAccount** [comment]
The existing collisions idea is actually pretty good - it hashes the soft bodies into a grid of buckets and only checks against the neighbouring buckets. The idea itself is far better than brute force and just as good as a QuadTree would be. A QuadTree is only needed if the space is very sparse or the soft bodies are different sizes, neither of which apply to evolvio - yet.

The problem is that the implementation is inefficient and slow eg:
1. It checks a bucket against all it's neighbours which makes it check every bucket pair twice and causes every collision to be counted twice
2. It re-creates the colliders list for each soft body rather than for each bucket, repeating the same work if multiple soft bodies are in the same bucket
3. It re-creates the colliders list every time rather than just updating the list when a soft body moves

If those issues can be solved it'll give a large performance improvement

**Comment by Holkin** [comment]
Funny thing, I tried to do it better but returned to original idea of splitting area into chunks, which allows handling collisions inside chunks, not whole board. This approach is used in particles modeling, but in our case one body could cover several chunks. 
Another important thing I should say, we must abandon OOP in code that directly affects performance. We should allocate all data into arrays, which allows CPU to use cache and reduce number of calls to RAM. But I'm not sure I have good solution about how this data structure should look like. And the most urgent and important issue is whether we should recalculate this data from scratch every frame or modify existing structure.
But I believe we don't have good algorithm at first. Just splitting list into set of sets, and then O(n2) inside this subsets looks not very good to me.

**Comment by Holkin** [comment]
I did some calculations. 
Simple idea of calculating collisions for whole board is about O(n2).
If we split it into chunks it will be O( (n*f(m))2 * m), where m - number of chunks, f(m) - number of chunks covered by one body. Let f(m) = a*m, where 0 < a < 1. This gives us O( a2*m*n2) = O(n2).
The same complexity, but with some coefficient. We can improve this coefficient by parallel processing, but still it's O(n2).
We need O (n*log(n)) at least. O(n) would be perfect.

**Comment by NateUniAccount** [comment]
Putting everything into arrays would be premature optimisation IMO, it might save a percentage point or two of performance but cause hours of delays and countless headaches for people trying to understand what data goes where and figuring out runtime bugs that could be prevented at compile time with a decent class structure. IMO it's just not worth that cost.

There may be some misunderstanding about the collision algorithm being used and it's Big O. The algorithm being used isn't brute force O(n^2), it already is a hashing algorithm (the chunking method you mention) which runs in O(c*n) = O(n) where c is a constant. In a good implementation that constant is small and performance is ok. The trouble is that the current implementation is causing the constant to be quite large (for the 3 reasons mentioned).

Multi-threading the calculations is an interesting idea but with the current scale the overhead of splitting jobs, ensuring there's no race conditions, joining threads etc. would IMO smother any speed increase it gave. If it were millions of creatures then yeah multi-threading would be a good way to go.

**Comment by UnknownEvil** [comment]
Here's some issues with the project as I've seen it.

The brain's neural network is a 2d array, BUT it's not jagged so each layer is the same size, which could be a problem in learning.

When I did inherited eye length and angle update I switched the places of 2 things in the NN inputs so those are mislabeled.

Inheritance has no random factors, that's a small problem. Perhaps creatures should inherit a mutability factor that is how far they can mutate from their parents (this should mutate generation to generation too)

It appears collision takes a while. So does calculating each eye for each creature, putting 50 eyes on a creature significantly slows down the game compared to having 10.

Configuration needs tweaking.

Attacking should immediately transfer the food rather than dropping it, I doubt the creatures would be capable of learning to eat from the tiles it drops on.

More issues with eyes, when it decides eye length it's not based on the max eye length, it's based on a constant (2), and since you can go no higher than your parents value since it's just the average and there's no mutation the eye length can't increase to the maximum. Random based on 0, max length would be better.

The UI is just plain unattractive and unintuitive.

We will need to do some optimizations because at the moment it runs pretty slow compared to what I think it could potentially run at.

(Will add more later)

**Comment by NateUniAccount** [comment]
What do you guys think, is it worth continuing with it and improving it bit by bit or spending a bit of time getting ideas together and designing something new? Or maybe even do both and have 2 separate repos and be able to swap ideas? I genuinely don't know at the moment.

**Comment by UnknownEvil** [comment]
If we do start over, would it be preferrable to switch to some other language aswell like C#?

**Comment by PRabahy** [comment]
If we are discussing a complete do over, I believe somebody already did a line-by-line port to Unity. https://github.com/evolvio/evolv.io/issues/88 Personally, I like C# more than Java, but I don't have a strong preference either way.

**Comment by NateUniAccount** [comment]
I'd be happy with both C# or Java.

It feels like going with the Unity repo would lead to the same problems we're seeing here because it's the same starting point. Unity itself may be a good choice though, or another tech depending on what everyone wants. It would be a completely from scratch project, new design, re-visit the algorithms and UI, see whether we could incorporate any new ideas anyone had. Once we're settled on what we want, split the design up so we can each do things in parallel. Make sure everything's nice and modular so it's easy for people to try different things with it, all that good stuff.

I think it would take more time to begin with, but once it's done and all modularised with interfaces, it'll be a lot quicker to try experiments and plug different things in like new brain designs or different ways of evolving.

**Comment by NateUniAccount** [comment]
So imagine we start over...what would we want to do? What techs or algorithms or ideas would we want to put in a new project?

To start:
Investigate QuadTree vs Hashing
2D vs 3D, or even options for both
Revisit how the Brain works, make it pluggable
Gather ideas for senses and how they could be implemented
Revisit how evolution, inheritance, mutation works
Revisit how creature interaction works
Potential for different species and how speciation would work
Look at what makes a creature, it's size, body parts, maybe pluggable body parts so a body phenotype could be built from a random genotype

**Comment by Holkin** [comment]
Why? Just start it simple and see how it goes.

**Comment by oscardssmith** [comment]
Wrt collisions, one easy option is simply sorting by x, bucketing, and sorting each bucket by y. Although it's technically an O(n^2) algorithm, almost all of the work is done with O(n log(n)) sorting that is very fast, so even though the theoretical runtime isn't great, it should be very fast.

**Comment by Holkin** [comment]
I have O(n) :)

**Comment by Holkin** [comment]
Like solar eclipse. :)
You can see how creature #19 passes through creature #6.

get collisions start
get collisions return 0
get collisions start
new collision 19 6
get collisions return 1
get collisions start
new collision 19 6
new collision 19 6
get collisions return 1
get collisions start
new collision 19 6
new collision 19 6
get collisions return 1
get collisions start
new collision 19 6
new collision 19 6
get collisions return 1
get collisions start
new collision 19 6
new collision 19 6
get collisions return 1
get collisions start
new collision 19 6
new collision 19 6
get collisions return 1
get collisions start
new collision 19 6
new collision 19 6
get collisions return 1
get collisions start
new collision 19 6
new collision 19 6
get collisions return 1
get collisions start
new collision 19 6
new collision 19 6
new collision 19 6
get collisions return 1
get collisions start
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
get collisions return 1
get collisions start
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
get collisions return 1
get collisions start
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
get collisions return 1
get collisions start
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
get collisions return 1
get collisions start
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
get collisions return 1
get collisions start
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
get collisions return 1
get collisions start
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
get collisions return 1
get collisions start
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
get collisions return 1
get collisions start
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
get collisions return 1
get collisions start
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
get collisions return 1
get collisions start
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
get collisions return 1
get collisions start
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
get collisions return 1
get collisions start
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
get collisions return 1
get collisions start
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
get collisions return 1
get collisions start
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
get collisions return 1
get collisions start
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
get collisions return 1
get collisions start
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
get collisions return 1
get collisions start
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
get collisions return 1
get collisions start
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
new collision 19 6
get collisions return 1
get collisions start
new collision 19 6
new collision 19 6
new collision 19 6
get collisions return 1
get collisions start
new collision 19 6
new collision 19 6
new collision 19 6
get collisions return 1
get collisions start
new collision 19 6
new collision 19 6
new collision 19 6
get collisions return 1
get collisions start
new collision 19 6
new collision 19 6
new collision 19 6
get collisions return 1
get collisions start
new collision 19 6
new collision 19 6
get collisions return 1
get collisions start
new collision 19 6
new collision 19 6
get collisions return 1
get collisions start
get collisions return 0

**Comment by Holkin** [comment]
```
    private static void draw(ISoftBody body, int[][] m, Set<Collision> set) {
        int offx = (int) Math.round(body.getPx() * World.RESOLUTION);
        int offy = (int) Math.round(body.getPy() * World.RESOLUTION);
        int r = (int) Math.round(body.getRadius() * World.RESOLUTION);
        int id = body.getId();

        double r2 = (r + .5) * (r + .5);
        for (int i = 0; i <= r; i++) {
            int x = (int) (Math.sqrt(r2 - i * i));
            for (int j = offx - x; j <= offx + x; j++) {
                try {
                    if (m[offy + i][j] > -1) {
                        set.add(new Collision(body, Singletons.FAUNA.getById(m[offy + i][j])));
                        System.out.printf("new collision %d %d\n", id, m[offy + i][j]);
                    }
                    m[offy + i][j] = id;
                    if (i == 0) continue;
                    if (m[offy - i][j] > -1) {
                        set.add(new Collision(body, Singletons.FAUNA.getById(m[offy - i][j])));
                        System.out.printf("new collision %d %d\n", id, m[offy - i][j]);
                    }
                    m[offy - i][j] = id;
                } catch (ArrayIndexOutOfBoundsException ex) {
                    continue; // it's ok
                }
            }
        }
    }

    private static void clear(int[][] m) {
        for (int i = 0; i < m.length; i++) {
            for (int j = 0; j < m[i].length; j++) {
                m[i][j] = -1;
            }
        }
    }
```

**Comment by UnknownEvil** [comment]
I have no idea what's going on here with the Os and the ns

**Comment by NateUniAccount** [comment]
O(n) etc. is the "Big O" of an algorithm, it's the expected running time dependent on n

**Comment by Holkin** [comment]
This is a function that shows how fast calculation cost grows.

**Comment by DiZZ3** [comment]
I am by no means a software dev but i have recently been self teaching myself how to program in c#. I would love to contribute on this project and help out where i can in this. **I suggest we get the ball rolling and open a separate repo for all of this.** I dont have enough experience with github to get this done but whoever wants to take lead go ahead and create one and link it here.  Also, I do have a little experience with unity and i feel like its a great place to start.

**Comment by UnknownEvil** [comment]
I don't know if we should use Unity though.

**Comment by NateUniAccount** [comment]
We could have 1 main holder project with a couple of sub repos, say one for Unity, one for Processing, one for plain Java. It'll keep it all in one place and we'd be able to easily see how the other projects are solving problems.

**Comment by UnknownEvil** [comment]
I like this idea

**Comment by DiZZ3** [comment]
Me too.

**Comment by NateUniAccount** [comment]
OK I tried adding a new repo to this evolvio but it looks like the evolvio user or account owner needs to do it, anyone know who that is?

**Comment by PRabahy** [comment]
I think https://github.com/mbertheau is. He was the first person to commit.

**Comment by UnknownEvil** [comment]
Can anyone get in touch with them so we can get this rolling?

**Comment by NateUniAccount** [comment]
Looks like he's given up on it, will create a new account for it and link it here

**Comment by evolvio4all** [comment]
Ok there's a new account here

https://github.com/evolvio4all

It's got 3 bare repos in it to begin with, if anyone wants another one just ask :)

**Comment by Holkin** [comment]
What are we going to start with?

**Comment by UnknownEvil** [comment]
Alright, so are we starting from scratch then?

**Comment by NateUniAccount** [comment]
I'm gonna be working on the Java one, probably do a skeleton game loop and blank ui to begin with, then flesh it out from there

**Comment by NateUniAccount** [comment]
I'm gonna be working on the Java one, probably do a skeleton game loop and blank ui to begin with, then flesh it out from there

Sure will add a JavaScript one!

**Comment by NateUniAccount** [comment]
JavaScript repo added

**Comment by UnknownEvil** [comment]
Thanks, I'll start work on it as soon as I get home.

**Comment by UnknownEvil** [comment]
So, should we put the new link in the README so people can find it?

**Comment by quete12** [comment]
I Started a complete overhaul of this project about 2 weeks ago but I messed up the UI. I want to share the refactored/ recreated code. I'd be happy to see that link in a Readme. I've now started to create a new project like this one using lwjgl because this looks like a solid base and gives me some useful insights for the job I'm looking for.

**Comment by quete12** [comment]
I just uploaded commited the code to https://github.com/evolvio4all/evolvio4Processing/pull/1/commits/8ead5cff3c4192966a2ed1f0650950748fe523f3

**IMPORTANT!** The Code compiles as is but it needs some tweeking in the UI and how the Brain recieves it's inputs.

**Comment by NateUniAccount** [comment]
The idea for the new repos is to start the projects from scratch rather than keep working on the original codebase. Feel free to open PRs for here though
Side note: In general PRs usually get committed much quicker if they're small changes that do one thing. It may not even be possible/feasible to commit a PR with massive changes across many files as other people will still need to work on other changes at the same time

**Comment by NateUniAccount** [comment]
Good idea about the readme link
